src/palimpzest/operators/physical.py

```diff
@@ -825,6 +825,116 @@ def _passesFilter(self, candidate):
+        return state + 1
+    elif (func.lower() == 'average'):
+        sum, cnt = state
+        return (sum + val, cnt + 1)
+    else:
+        raise Exception("Unknown agg function " + func)
+
+def agg_final(func, state):
+    if (func.lower() == 'count'):
+        return state
+    elif (func.lower() == 'average'):
+        sum, cnt = state
+        return float(sum)/cnt
+    else:
+        raise Exception("Unknown agg function " + func)
+
+
+class ApplyGroupByOp(PhysicalOp):
+        def __init__(self, source: PhysicalOp, gbySig: GroupBySig,  targetCacheId: str=None, shouldProfile=False):
+            super().__init__(outputSchema=gbySig.outputSchema(), shouldProfile=shouldProfile)
+            self.source = source
+            self.gbySig = gbySig
+            self.targetCacheId=targetCacheId
+            self.shouldProfile=shouldProfile
+
+        def __str__(self):
+            return str(self.gbySig)
+        def opId(self):
+            d = {
+                "operator": "ApplyGroupByOp",
+                "source": self.source.opId(),
+                "gbySig": str(GroupBySig.serialize(self.gbySig)),
+                "targetCacheId": self.targetCacheId,
+            }
+            ordered = json.dumps(d, sort_keys=True)
+            return hashlib.sha256(ordered.encode()).hexdigest()[:MAX_ID_CHARS]
+        def dumpPhysicalTree(self):
+            """Return the physical tree of operators."""
+            return (self, self.source.dumpPhysicalTree())
+        
+        def estimateCost(self):
+            inputEstimates = self.source.estimateCost()
+
+            outputEstimates = {**inputEstimates}
+            outputEstimates['cardinality'] = 1
+
+            # for now, assume applying the aggregate takes negligible additional time (and no cost in USD)
+            outputEstimates['timePerElement'] = 0
+            outputEstimates['usdPerElement'] = 0
+            outputEstimates['estOutputTokensPerElement'] = 0
+
+            return outputEstimates
+
+
+
+        def __iter__(self):
+            datadir = DataDirectory()
+            shouldCache = datadir.openCache(self.targetCacheId)
+            aggState = {}
+
+            @self.profile(name="gby", op_id=self.opId(), shouldProfile=self.shouldProfile)
+            def iteratorFn():
+                for r in self.source:
+                    #build group array
+                    group = ()
+                    for f in self.gbySig.gbyFields:
+                        group = group + (getattr(r,f),)
+                    if group in aggState:
+                        state = aggState[group]
+                    else:
+                        state = []
+                        for fun in self.gbySig.aggFuncs:
+                            state.append(agg_init(fun))
+                    for i in range(0,len(self.gbySig.aggFuncs)):
+                        fun = self.gbySig.aggFuncs[i]
+                        field = getattr(r, self.gbySig.aggFields[i])
```

Comment:
If the GroupBySig is not correctly defined, then I think this is the line on which we would expect to see a problem.  The getAttr() will fail here.  Maybe we can have more explicit field compatibility checking so the user gets a more informative error message?

Reply:
Yes agreed, some error checking here would be good.